examples/omulti.ml

```diff
@@ -16,73 +16,67 @@ let finished mt =
-    let event = match flags with
-                | Ev.READ -> M.EV_IN
-                | Ev.WRITE -> M.EV_OUT
-                | _ -> M.EV_AUTO
-    in
-    let _ = M.action mt fd event in
-    finished mt
-  in
-  let evs = ref 0 in
-  M.set_socket_function mt begin fun fd what ->
-    List.iter (fun ev -> decr evs; Ev.del ev) (Hashtbl.find_all events fd); Hashtbl.remove events fd;
-    let flags = match what with
-          | M.POLL_REMOVE | M.POLL_NONE -> []
-          | M.POLL_IN -> [Ev.READ]
-          | M.POLL_OUT -> [Ev.WRITE]
-          | M.POLL_INOUT -> [Ev.READ;Ev.WRITE]
-    in
-    match flags with
-    | [] -> finished mt
-    | flags ->
-      let ev = Ev.create () in
-      Ev.set events_base ev fd flags ~persist:true on_event;
-      Ev.add ev None;
-      incr evs;
-      Hashtbl.add events fd ev
-  end;
-  let _ = M.action_all mt in
-  Ev.dispatch events_base;
-  assert (0 = !evs);
-  finished mt
-
-let loop_select mt =
+let loop_select hs mt =
   pr "action/select";
   let in_fd = ref [] in
   let out_fd = ref [] in
+  let is_done = ref (A.length hs = 0) in
+  let m = Mutex.create () in
   let on_event fd event =
-    let _ = M.action mt fd event in
-    finished mt
+    if M.action mt fd event = 0 then
+      mutexify m (fun () ->
+        is_done := true
+      ) ();
   in
-  M.set_socket_function mt begin fun fd what ->
-    in_fd := List.filter ((<>) fd) !in_fd;
-    out_fd := List.filter ((<>) fd) !out_fd;
-    match what with
-    | M.POLL_REMOVE | M.POLL_NONE -> finished mt
-    | M.POLL_IN -> in_fd := fd :: !in_fd
-    | M.POLL_OUT -> out_fd := fd :: !out_fd
-    | M.POLL_INOUT -> in_fd := fd :: !in_fd; out_fd := fd :: !out_fd
-  end;
-  let _ = M.action_all mt in
-  while !in_fd <> [] || !out_fd <> [] do
-    let (fdi,fdo,_) = Unix.select !in_fd !out_fd [] (-1.) in
-    List.iter (fun fd -> on_event fd M.EV_IN) fdi;
-    List.iter (fun fd -> on_event fd M.EV_OUT) fdo;
-  done;
+  M.set_socket_function mt (fun fd what -> 
+    A.iter (fun (h, url) ->
+      if Some fd = Curl.get_activesocket h then
+        pr "Waiting for event on socket for url %s" url) hs;
+    mutexify m (fun () ->
+      in_fd := List.filter ((<>) fd) !in_fd;
+      out_fd := List.filter ((<>) fd) !out_fd;
+      (match what with
+         | M.POLL_REMOVE | M.POLL_NONE -> ()
+         | M.POLL_IN -> in_fd := fd :: !in_fd
+         | M.POLL_OUT -> out_fd := fd :: !out_fd
+         | M.POLL_INOUT -> in_fd := fd :: !in_fd; out_fd := fd :: !out_fd)) ()
+  );
+  let rec process () =
+    M.action_timeout mt;
+    let is_done = 
+      mutexify m (fun () ->
+        !is_done
+      ) ()
+    in
+    if not is_done then (
+      let timeout =
+        if !in_fd = [] && !out_fd = [] then 0.1 else -1.
+      in
+      let (fdi,fdo,_) = Unix.select !in_fd !out_fd [] timeout in
+      List.iter (fun fd -> on_event fd M.EV_IN) fdi;
+      List.iter (fun fd -> on_event fd M.EV_OUT) fdo;
+      process ()
+    )
+  in
+  let th = Thread.create process () in
```

Comment:
thread is not needed?

Reply:
and all the mutexes either? all the callbacks are invoked in the context of select, so it is always same thread, there is no concurrent code running

Reply:
actually only now I understand that adding this thread stuff is the only essential change in omulti.ml? Please don't change formatting.